core-rust/state-manager/src/store/historical_state.rs

```diff
@@ -0,0 +1,657 @@
+struct TreeStoreBrowser<T> {
+    tree_store: T,
+    version: Version,
+}
+
+impl<'t, T: Deref<Target = impl ReadableTreeStore> + Clone + 't> TreeStoreBrowser<T> {
+    /// Creates an instance directly.
+    pub fn new(tree_store: T, version: Version) -> Self {
+        Self {
+            tree_store,
+            version,
+        }
+    }
+
+    /// Creates an instance for browsing a [`NestedTreeStore`] nested at the given leaf.
+    /// This effectively allows to access a lower Tier tree (of a Radix-specific 3-Tier JMT).
+    pub fn nested(&self, leaf: &ResolvedLeaf) -> TreeStoreBrowser<Rc<NestedTreeStore<T>>> {
+        let nested_tree_store =
+            NestedTreeStore::new(self.tree_store.clone(), leaf.nibble_path.bytes().to_vec());
+        TreeStoreBrowser::new(Rc::new(nested_tree_store), leaf.last_hash_change_version)
+    }
+
+    /// Returns a specific leaf, if found by starting at the scoped version's root and following the
+    /// given [`NibblePath`].
+    pub fn get_leaf(&self, nibble_path: &NibblePath) -> Option<ResolvedLeaf> {
+        self.iter_leafs_from(nibble_path)
+            .next()
+            .filter(|leaf| &leaf.nibble_path == nibble_path)
+    }
+
+    /// Returns an iterator of all leafs reachable from the scoped version's root, in
+    /// lexicographical order, starting from the given one.
+    pub fn iter_leafs_from(
+        &self,
+        nibble_path: &NibblePath,
+    ) -> Box<dyn Iterator<Item = ResolvedLeaf> + 't> {
+        recurse_until_leafs(
+            self.tree_store.clone(),
+            NodeKey::new_empty_path(self.version),
+            nibble_path.nibbles().collect(),
+        )
+    }
+}
+
+/// Returns a lexicographically-sorted iterator of all the `tree_store`'s [`ResolvedLeaf`]s having
+/// [`NibblePath`]s greater or equal to the given `from_nibbles`.
+///
+/// The algorithm:
+/// - starts at the given `at_key`,
+/// - then goes down the tree, guided by the given `from_nibbles` chain, for as long as it is
+///   possible,
+///   - Note: this means it will either locate exactly this nibble path, or - if it does not
+///     exist - settle at its direct successor.
+/// - and then continues as if it was a classic DFS all the way,
+/// - but only leaf nodes are returned.
+///
+/// The implementation is a lazy recursive composite of child iterators.
+fn recurse_until_leafs<'t, T: Deref<Target = impl ReadableTreeStore> + Clone + 't>(
+    tree_store: T,
+    at_key: NodeKey,
+    from_nibbles: VecDeque<Nibble>,
+) -> Box<dyn Iterator<Item = ResolvedLeaf> + 't> {
+    let Some(node) = tree_store.get_node(&at_key) else {
+        panic!("{:?} referenced but not found in the storage", at_key);
+    };
+    match node {
+        TreeNode::Internal(internal) => {
+            let mut child_from_nibbles = from_nibbles;
+            let from_nibble = child_from_nibbles
+                .pop_front()
+                .unwrap_or_else(|| Nibble::from(0));
```

Comment:
Super minor comment, but my slight preference would be to use an Option here (and e.g. `from_nibble.map_or(true, |fn| child.nibble >= fn)` below).
I know this totally doesn't matter (it's just my OCD :smile: ) as both cases are really the same, but in the `if ` below we could go the `mem:take` path if we're actually comparing against the last `0` from `from_nibbles` and  `VecDeque::new()` if it's None (or a "default 0"). Currently we go the mem:take path in both cases.
But again, that's just a personal-preference-based nitpicking and doesn't influence the correctness, feel free to ignore.

Reply:
Done; it turns out that handling the Option here is actually even less code than the `Nibble::from(0)` 👍